File: 606-construct-string-from-binary-tree/606-construct-string-from-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    String ans = "";
    public String tree2str(TreeNode root) {
        if(root == null) {
            return "";
        }
        
        ans += root.val;
        if(root.left != null) {
            ans += '(';
            ans = tree2str(root.left);
            ans += ')';
        } else if(root.right != null) {
            ans += '(';
            ans += ')'; 
        }
        
        if(root.right != null) {
            ans += '(';
            ans = tree2str(root.right);
            ans += ')';
        }
        
        return ans;
    }
}

File: 606-construct-string-from-binary-tree/606-construct-string-from-binary-tree.cpp
/**
 *Definition for a binary tree node.
 *struct TreeNode {
 *    int val;
 *    TreeNode * left;
 *    TreeNode * right;
 *    TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 *};
 */
class Solution {
public:
    string ans = "";
    string tree2str(TreeNode *t) {
        if (!t)
            return "";

        ans += to_string(t->val);

        if (t->left) {
            ans += '(';
            ans = tree2str(t->left);
            ans += ')';
        }
        else if (t->right) {
            ans += '(';
            ans += ')';
        }

        if (t->right) {
            ans += '(';
            ans = tree2str(t->right);
            ans += ')';
        }

        return ans;
    }
};